src/Coordination/KeeperStorage.cpp

```diff
@@ -889,65 +1290,122 @@ struct KeeperStorageMultiRequestProcessor final : public KeeperStorageRequestPro
+        response_errors.reserve(concrete_requests.size());
+        for (size_t i = 0; i < concrete_requests.size(); ++i)
         {
-            size_t i = 0;
-            for (const auto & concrete_request : concrete_requests)
-            {
-                auto [ cur_response, undo_action ] = concrete_request->process(storage, zxid, session_id, time);
+            auto new_deltas = concrete_requests[i]->preprocess(storage, zxid, session_id, time);
 
-                response.responses[i] = cur_response;
-                if (cur_response->error != Coordination::Error::ZOK)
+            if (!new_deltas.empty())
+            {
+                if (auto * error = std::get_if<KeeperStorage::ErrorDelta>(&new_deltas.back().operation))
                 {
-                    for (size_t j = 0; j <= i; ++j)
-                    {
-                        auto response_error = response.responses[j]->error;
-                        response.responses[j] = std::make_shared<Coordination::ZooKeeperErrorResponse>();
-                        response.responses[j]->error = response_error;
-                    }
+                    std::erase_if(saved_deltas, [zxid](const auto & delta) { return delta.zxid == zxid; });
+
+                    response_errors.push_back(error->error);
 
-                    for (size_t j = i + 1; j < response.responses.size(); ++j)
+                    for (size_t j = i + 1; j < concrete_requests.size(); ++j)
                     {
-                        response.responses[j] = std::make_shared<Coordination::ZooKeeperErrorResponse>();
-                        response.responses[j]->error = Coordination::Error::ZRUNTIMEINCONSISTENCY;
+                        response_errors.push_back(Coordination::Error::ZRUNTIMEINCONSISTENCY);
                     }
 
-                    for (auto it = undo_actions.rbegin(); it != undo_actions.rend(); ++it)
-                        if (*it)
-                            (*it)();
+                    return {{zxid, KeeperStorage::FailedMultiDelta{std::move(response_errors)}}};
+                }
+            }
+            new_deltas.emplace_back(zxid, KeeperStorage::SubDeltaEnd{});
+            response_errors.push_back(Coordination::Error::ZOK);
+
+            saved_deltas.insert(saved_deltas.end(), std::make_move_iterator(new_deltas.begin()), std::make_move_iterator(new_deltas.end()));
+        }
+
+        return {};
+    }
+
+    Coordination::ZooKeeperResponsePtr process(KeeperStorage & storage, int64_t zxid, int64_t session_id, int64_t time) const override
+    {
+        Coordination::ZooKeeperResponsePtr response_ptr = zk_request->makeResponse();
+        Coordination::ZooKeeperMultiResponse & response = dynamic_cast<Coordination::ZooKeeperMultiResponse &>(*response_ptr);
+
+        auto & deltas = storage.uncommitted_state.deltas;
+        if (auto * failed_multi = std::get_if<KeeperStorage::FailedMultiDelta>(&deltas.front().operation))
```

Comment:
It's unclear how we verify that `deltas.front()` contains exactly the delta for this request. Maybe add additional checks to make invariants clear.